### ISSUE TITLE
Feature Request: EXCLUDE MATCH

### ISSUE THREAD

**Issue by InverseFalcon** [neo4j/neo4j#7762]
This proposes an EXCLUDE MATCH clause, which must follow a WITH, as it has no meaning when not using variables from a previous MATCH. After the EXCLUDE MATCH, the in-scope variables should be the same ones that were piped in with the last WITH.

The purpose is to look for pattern match using previously defined variables, and then remove any rows where that pattern occurs. 

This behavior can be replicated currently in Cypher by looking for a pattern with OPTIONAL MATCH, WHERE the resulting pattern is null.

An example of usage - in a graph of :Patients and :Doctors (where :Doctors may work with other :Doctors), I'm looking for :Patients whose :Doctor (or any Doctor they are associated with through chained :WorksWith relationships, however distantly) is not Doctor Frankenstein or Doctor Claw.

In current Cypher I might use this:

```
WITH ['Frankenstein', 'Claw'] as badDoctors
MATCH (patient:Patient)-[:TreatedBy]->(doctor:Doctor)
WHERE doctor.name NOT IN badDoctors
WITH patient, doctor
OPTIONAL MATCH badDoctorPattern = (doctor)-[:WorksWith*]-(badDoctor:Doctor)
WHERE badDoctor.name IN badDoctors
WITH patient, doctor, badDoctorPattern
WHERE badDoctorPattern IS NULL
RETURN patient
```

With the new EXCLUDE MATCH clause, we don't save too many lines, but it should make the intent of the query more clear:

```
WITH ['Frankenstein', 'Claw'] as badDoctors
MATCH (patient:Patient)-[:TreatedBy]->(doctor:Doctor)
WHERE doctor.name NOT IN badDoctors
WITH patient, doctor
EXCLUDE MATCH (doctor)-[:WorksWith*]-(badDoctor:Doctor)
WHERE badDoctor.name IN badDoctors
RETURN patient
```

The above example could of course be optimized to match on the bad :Doctors first, then use WHERE NOT EXISTS() and supply the badDoctor pattern, avoiding the need for any kind of match. 

While that circumvents the need for an EXCLUDES MATCH here, there are other cases where a pattern alone is not rich enough to express what's needed (maybe there needs to be a predicate on the relationship, or some other condition where a predicate needs to be present). EXCLUDES MATCH lets us tackle queries where NOT EXISTS(pattern) isn't rich enough for our needs, and where the OPTIONAL MATCH workaround obfuscates the intent of the query.

**Comment by jexp** [comment]
Is it the same as?

WITH ['Frankenstein', 'Claw'] as badDoctors
MATCH (patient:Patient)-[:TreatedBy]->(doctor:Doctor)
WHERE doctor.name NOT IN badDoctors
AND NOT EXISTS ( (doctor)-[:WorksWith*]-(badDoctor:Doctor) )
WHERE badDoctor.name IN badDoctors
RETURN patient

Which already works.

**Comment by InverseFalcon** [comment]
Does that work though? I'm getting a syntax error due to the two WHERE clauses being back to back. I'm under the impression that WHERE can only be associated with START, MATCH, OPTIONAL MATCH, and WITH, and only once, not multiple times.

If this can be made to work, if I'm doing something wrong, then all is good. One of the reasons EXLUDE MATCH seemed needed was I couldn't seem to use another WHERE upon the previous WHERE clause to add predicates on a pattern in NOT EXISTS.

**Comment by jexp** [comment]
Sorry WHERE ... AND ...

Von meinem iPhone gesendet

**Comment by InverseFalcon** [comment]
I get a variable not defined error (on badDoctor). At least in 3.0.3, I don't believe you can introduce new variables within the EXISTS pattern.

**Comment by sherfert** [comment]
Would matching the bad doctor before the EXISTS solve your issue? Like in

```
WITH ['Frankenstein', 'Claw'] as badDoctors
MATCH (patient:Patient)-[:TreatedBy]->(doctor:Doctor), (badDoctor:Doctor)
WHERE NOT doctor.name IN badDoctors
AND badDoctor.name IN badDoctors
AND NOT EXISTS ( (doctor)-[:WorksWith*]-(badDoctor) )
RETURN patient
```

**Comment by craigtaverner** [comment]
To avoid the cartesian product, I would do something like:

```
WITH ['Frankenstein', 'Claw'] as badDoctorNames
MATCH (badDoctor:Doctor) WHERE badDoctor.name IN badDoctorNames
WITH collect(badDoctor) badDoctors
MATCH (patient:Patient)-[:TreatedBy]->(doctor:Doctor)
WHERE NOT doctor.name IN badDoctorNames
AND NONE(badDoctor IN badDoctors WHERE  (doctor)-[:WorksWith*]-(badDoctor) )
RETURN patient
```

**Comment by InverseFalcon** [comment]
Yes, those work fine. It would still be nice to have some syntactical sugar with a EXCLUDES MATCH or a NOT EXISTS( <pattern> WHERE)

**Comment by sherfert** [comment]
Has been moved to openCypher: https://github.com/opencypher/openCypher/issues/258
